internal/proxy/proxy.go

```diff
@@ -60,38 +61,113 @@ type Client struct {
 	mnts []*socketMount
 }
 
-// NewClient completes the initial setup required to get the proxy to a "steady" state.
-func NewClient(ctx context.Context, cmd *cobra.Command, conf *Config) (*Client, error) {
-	d, err := cloudsqlconn.NewDialer(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("error initializing dialer: %v", err)
+type portConfig struct {
+	mu        sync.Mutex
+	global    int
+	postgres  int
+	mysql     int
+	sqlserver int
+}
+
+func newPortConfig(global int) *portConfig {
+	return &portConfig{
+		global:    global,
+		postgres:  5432,
+		mysql:     3306,
+		sqlserver: 1433,
+	}
+}
+
+// nextPort returns the next port based on the initial global value.
+func (c *portConfig) nextPort() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	p := c.global
+	c.global++
+	return p
+}
+
+func (c *portConfig) nextDBPort(version string) int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	switch {
+	case strings.HasPrefix(version, "MYSQL"):
+		p := c.mysql
+		c.mysql++
+		return p
+	case strings.HasPrefix(version, "POSTGRES"):
+		p := c.postgres
+		c.postgres++
+		return p
+	case strings.HasPrefix(version, "SQLSERVER"):
+		p := c.sqlserver
+		c.sqlserver++
+		return p
+	default:
+		// Unexpected engine version, use global port setting instead.
+		return c.nextPort()
 	}
-	c := &Client{cmd: cmd, dialer: d}
+}
 
-	port := conf.Port
-	var inc int
+// NewClient completes the initial setup required to get the proxy to a "steady" state.
+func NewClient(ctx context.Context, d *cloudsqlconn.Dialer, cmd *cobra.Command, conf *Config) (*Client, error) {
+	pc := newPortConfig(conf.Port)
+	type mountRes struct {
+		mnt *socketMount
+		err error
+	}
+	mntCh := make(chan mountRes, len(conf.Instances))
 	for _, inst := range conf.Instances {
-		m := &socketMount{inst: inst.Name}
-		a := conf.Addr
-		if inst.Addr != "" {
-			a = inst.Addr
-		}
-		nextPort := port + inc
-		if inst.Port != 0 {
-			nextPort = inst.Port
-		} else {
-			inc++
+		go func(i InstanceConnConfig) {
+			m := &socketMount{inst: i.Name}
+			a := conf.Addr
+			if i.Addr != "" {
+				a = i.Addr
+			}
```

Comment:
nit: Is it possible to set this when the query string is parsed? Something like "if not ok, set default value"?

Reply:
As in, when we don't see an address query string, we should just parse the instance's address as the default value? I think that would remove the need for a top level address field.

Reply:
I'd like to hold off on this change for now. We're checking the port's global config vs instance config in this same function, and for now keeping this similar code together seems preferable, but I could be wrong.

Reply:
SGTM, we can punt this until later.